storage/block/block.go

```diff
@@ -194,26 +193,28 @@ func (b *dbBlock) Stream(blocker context.Context) (xio.SegmentReader, error) {
 		// NB(r): We explicitly create a new segment to ensure references
 		// are taken to the bytes refs and to not finalize the bytes.
 		stream.Reset(ts.NewSegment(b.segment.Head, b.segment.Tail, ts.FinalizeNone))
+		blocker.RegisterFinalizer(stream)
 	}
 
-	if b.mergeTarget != nil {
-		var mergeStream xio.SegmentReader
-		mergeStream, err = b.mergeTarget.Stream(blocker)
-		if err != nil {
-			stream.Finalize()
-			return nil, err
-		}
-		// Return a lazily merged stream
-		// TODO(r): once merged reset this block with the contents of it
-		stream = newDatabaseMergedBlockReader(b.startWithLock(), stream, mergeStream, b.opts)
+	if b.mergeTarget == nil {
+		return stream, nil
 	}
 
-	if !fromBlockRetriever {
-		// Register the finalizer for the stream, the block retriever already
-		// registers the stream as a finalizer for the context so we only perform
-		// this if we return a stream directly from this block
-		blocker.RegisterFinalizer(stream)
+	var mergeStream xio.SegmentReader
```

Comment:
+1 for the order refactor, the early termination makes the branches much easier to reason about